[WIP] initial implementation to support audio processing as arrays

g8a9 opened the pull request:
This PR has multiple goals. 
The main one is to **support using numpy arrays for audio rather than local paths only.** I believe this could be a strong limitation since many audio datasets today come from the HF Hub already decoded into np.ndarray.

However, while inspecting the code, I've noticed there are still **several dependencies from local files that we don't want to have.** For example, the code expects:
- local transcripts for most of the explainers (I think we should have a centralized place where we compute them)
- local files for adding noise (white and pink)

Another important goal is to **standardize transcript generation,** which is currently made only by the LOO explainers.

g8a9 committed:
[WIP] initial implementation to support audio processing as arrays

g8a9 commented:
@emanuele-moscato I've left around several todo markers for points to discuss. You can find them with `#TODO GA`

emanuele-moscato commented:
Look at the explainers under `ferret/explainers/explanation_speech/`:

- [x] Whenever the path to an audio file is required (typically in the `compute_explanation` methods), pass an object of type `FerretAudio` (newly defined in the PR).
- [x] Make `FerretAudio` able to transcribe to text (dedicated method called from within the explainers when/if needed). --> NOT VALID ANYMORE
- [x] Check the SpeechXAI examples notebook.

gaiageagea committed:
proposed changes to integrate numpy arrays

emanuele-moscato commented on ferret/explainers/explanation_speech/gradient_speech_explainer.py:
```diff
@@ -181,7 +177,7 @@ def compute_explanation(
             explainer=self.NAME + "-" + aggregation,
             target=targets if n_labels > 1 else targets,
-            audio_path=audio_path,
+            audio_path=audio,
```
Let's change the name of this optional argument in to `audio`, for consistency. This needs to be done in the `ExplanationSpeech` class within `explanation_speech.py`.

emanuele-moscato commented on ferret/explainers/explanation_speech/lime_speech_explainer.py:
```diff
@@ -143,7 +140,7 @@ def compute_explanation(
             explainer=self.NAME + "+" + removal_type,
             target=targets if n_labels > 1 else targets,
-            audio_path=audio_path,
+            audio_path=audio,
```
As in the other explainers: change keyword argument's name to `audio` in the `ExplanationSpeech` class.

emanuele-moscato commented on ferret/explainers/explanation_speech/loo_speech_explainer.py:
```diff
@@ -1,11 +1,15 @@
+from ...speechxai_utils import pydub_to_np, FerretAudio
+from logging import getLogger
+
+logger = getLogger(__name__)
```
We are starting to use loggers in different places: let's use a singleton pattern so we avoid having multiple ones instantiated at the same time.

emanuele-moscato commented on ferret/explainers/explanation_speech/loo_speech_explainer.py:
```diff
@@ -125,7 +127,7 @@ def compute_explanation(
             explainer=self.NAME + "+" + removal_type,
             target=targets if n_labels > 1 else [targets],
-            audio_path=audio_path,
+            audio=audio,  # TODO GA: I don't know if this is something we want to keep
```
As above, but the keyword argument's name here has been updated already, but now I'm confused: the class hasn't been modified!

emanuele-moscato commented on ferret/speechxai_utils.py:
```diff
@@ -7,6 +7,81 @@
+            return self.array
+
+    @property
+    def transcribe(self):
```
I would have called this `transcription` as it's used as an attribute - but I'm aware I'm being pedantic here ahah!

emanuele-moscato commented on ferret/speechxai_utils.py:
```diff
@@ -7,6 +7,81 @@
 from transformers import Wav2Vec2ForSequenceClassification, Wav2Vec2FeatureExtractor
+import librosa
+from typing import Union
+from explainers.explanation_speech.utils_removal import transcribe_audio
```
Watch out: if I try to import stuff from the current module (e.g. the `FerretAudio` class), I get an error because the `explainers` is not found: we need a relative import here (`from .explainers....`). But now `utils_removal.py` tries to import from `speechxai_utils.py`, generating a circular import.

Commit:
Attach console handler to logger, fix bugs, move audio transcription to the speechxai_utils.py module

emanuele-moscato commented:
To do:
- [x] In the definition of the `ExplanationSpeech` class, change the kwarg name from `audio_path` to `audio` and fix code accordingly everywhere the class is implemented (make sure an object of type `FerretAudio` is passed).
- [x] Address the comments in the last code review.
- [x] Complete the update of the speech explainers so they accept an object of type `FerretAudio` as their input.
- [x] See previous to-do list.

**Note:** the audio transcription functions `transcribe_audio` and `transcribe_audio_with_model` were moved from the `ferret/explainers/explanation_speech/utils_removal.py` module to `ferret/speechxai_utils.py` to avoid a circular import. The rest of the code has already been updated, but have one last check please!

gaiageagea committed:
implementing comments, using FerretAudio homogeneously across explainers, changes to ExplanationSpeech Class

emanuele-moscato commented on ferret/explainers/explanation_speech/equal_width/gradient_equal_width_explainer.py:
```diff
@@ -117,7 +113,7 @@ def compute_explanation(
             a, b = 0, 0  # 50, 20
 
-            duration_s = len(audio_as) / 1000
+            duration_s = len(audio_array) / audio.sample_rate # finds the duration from the array 
```
Let's double-check this: the original factor `1000` was converting fro milliseconds to seconds (so `audio_as` was assumed to be an array with lengths equal the duration in milliseconds) while dividing by the sample rate requires that `audio_array` is now as long as the number of samples for a frequency equals to the sample rate (in hz).

g8a9 replied:
@emanuele-moscato we should check this one, too

emanuele-moscato commented on ferret/explainers/explanation_speech/equal_width/lime_equal_width_explainer.py:
```diff
@@ -59,13 +56,13 @@ def compute_explanation(
         splits = []
 
-        duration_s = len(audio_as) / 1000
+        duration_s = len(audio_array) / audio.sample_rate # finds the duration from the array 
```
Same double-check as in `gradient_equal_width_explainer.py`.

emanuele-moscato commented on ferret/explainers/explanation_speech/equal_width/loo_equal_width_explainer.py:
```diff
@@ -65,19 +63,17 @@ def compute_explanation(
         audio_remove_segments = []
 
-        duration_s = len(audio) / 1000
+        duration_s = len(audio_array) / audio.sample_rate # finds the duration from the array 
```
Same double-check as in `gradient_equal_width_explainer.py`.

g8a9 replied:
the denominator `1000` assumed to work with milliseconds. Is this change still valid? @emanuele-moscato

Commit:
Adapt faithfulness measures for speech to use FerretAudio objects, fix notebook

emanuele-moscato commented:
Issues to solve:
- [x] When loading an audio file into a `FerretAudio` object and then extracting the `AudioSegment` object (`to_pydub` method), the audio gets distorted. Check `FerretAudio`'s `to_pydub` method.

**Action:** modified `FerretAudio`'s `to_pydub` method so as to create a pudyb `AudioSegment` object from an array that is always unnormalized and of dtype `int16`.

- [x] Move audio resampling (if the native sample rate is different from 16 KHz) to within the transcription method.
- [x] Move normalization of the array corresponding to the audio (if needed) to within the transcription method as well (so there are no ambiguities on whether the array provided by the user is normalized or not: it is what it is from the start).
- [x] Check that the new way of obtaining arrays from audio (`librosa.load`) returns the same as (`AudioSegment.load_wav` --> `pydub_to_np`).

**Result:** numerically it is, they both return an array of dtype float32 normalized by a factor 2 ** 15, but the shape is different as (for mono audio) `librosa` returns a flattened array while `pydub_to_np` returns an array of shape `(n_samples, 1)`.

- [x] Raise error if the numpy array or audio file passed to `FerretAudio` has more than one channel (**we only support mono audio!**).

**Action:** this is inferred looking at the shape of the array. 1-dimensional array --> good, 2-dimensional array --> good only if the trailing dimension is 1 (shape `(n_samples, 1)`), array of dimension > 2 --> shape not understood, raise exception.

- [x] Remove the `remove_word_np` function if it ends up not being used.

Commit:
Fix conversion of array to pydub  type

g8a9 committed:
- Add a new "transcribe" method
- if word timestamps are not provided they are generated on the fly
- each word timestamps expects word transcripts
- word timestamps are not external to the FerretAudio class
- add a new notebook to show this behavior

g8a9 committed:
- added some dependencies
- updated the new notebook
- adapted the paraling explainer
- [WIP] code crashed if no ffmpeg is found on the machine

g8a9 committed:
Everything works with no coding errors on the new notebooks.

- final edits to methods to update
- update the notebook name
- WIP need to check that everything returns expected results
- WIP need to check that the notebook with local loading works

g8a9 commented:
Regarding the normalization: we should not touch the input array unless 1) it's not normalized and 2) we are transcribing it with whisperX -- that's the only case where we really need normalization + 16kHz sampling rate.

g8a9 committed:
update readme

g8a9 committed:
update readme

g8a9 committed:
update readme

gaiageagea committed:
removing remove_word_np and its occurrences

gaiageagea committed:
changing equal width explainers to calculate the duration of the audio using the numpy array when the pydub AS is not needed

Commit:
Make sure to use normalized arrays in the evaluators for speech

Commit:
Fix notebook

Commit:
Update notebooks

Commit:
Iron out minor details (use normalized arrays in all the speech explainers), commit re-evaluated notebooks
